roqoqo-derive/src/operate_unitary_modes.rs

```diff
@@ -0,0 +1,150 @@
+            }
+        }
+    }
+}
+
+/// Dispatch to derive OperateSingleModeGate for enums and structs
+pub fn dispatch_struct_enum_single_mode_gate(input: DeriveInput) -> TokenStream {
+    let ident = input.ident;
+    match input.data {
+        Data::Struct(_ds) => single_mode_gate_struct(ident),
+        Data::Enum(de) => single_mode_gate_enum(de, ident),
+        _ => panic!("OperateSingleMode can only be derived on structs and enums"),
+    }
+}
+
+fn single_mode_gate_enum(de: DataEnum, ident: Ident) -> TokenStream {
+    let _variants_with_type = extract_variants_with_types(de).into_iter();
+    quote! {
+        #[automatically_derived]
+        impl OperateSingleModeGate for #ident{
+        }
+
+    }
+}
+
+fn single_mode_gate_struct(ident: Ident) -> TokenStream {
+    quote! {
+        #[automatically_derived]
+        impl OperateSingleModeGate for #ident{
+        }
+    }
+}
+
+pub fn dispatch_struct_enum_two_mode_gate(input: DeriveInput) -> TokenStream {
+    let ident = input.ident;
+    match input.data {
+        Data::Struct(_ds) => two_mode_gate_struct(ident),
+        Data::Enum(de) => two_mode_gate_enum(de, ident),
+        _ => panic!("OperateTwoModeGate can only be derived on structs and enums"),
+    }
+}
+
+fn two_mode_gate_enum(de: DataEnum, ident: Ident) -> TokenStream {
+    let _variants_with_type = extract_variants_with_types(de).into_iter();
+    quote! {
+        #[automatically_derived]
+        impl OperateTwoModeGate for #ident{
+        }
+    }
+}
+
+fn two_mode_gate_struct(ident: Ident) -> TokenStream {
+    quote! {
+        #[automatically_derived]
+        impl OperateTwoModeGate for #ident{
+        }
+    }
+}
+
+pub fn dispatch_struct_enum_multi_mode_gate(input: DeriveInput) -> TokenStream {
+    let ident = input.ident;
+    match input.data {
+        Data::Struct(_ds) => multi_mode_gate_struct(ident),
+        Data::Enum(de) => multi_mode_gate_enum(de, ident),
+        _ => panic!("OperateMultiModeGate can only be derived on structs and enums"),
+    }
+}
+
+fn multi_mode_gate_enum(de: DataEnum, ident: Ident) -> TokenStream {
+    let variants_with_type = extract_variants_with_types(de).into_iter();
+    let match_quotes = variants_with_type.map(|(vident, _, _)| {
+        quote! {
+            &#ident::#vident(ref inner) => {OperateMultiModeGate::circuit(&(*inner))},
+        }
+    });
+    quote! {
+        #[automatically_derived]
+        impl OperateMultiModeGate for #ident{
+            fn circuit(&self) -> crate::Circuit {
+                match self{
+                    #(#match_quotes)*
+                    _ => panic!("Unexpectedly cannot match variant")
+                }
+            }
+        }
+    }
+}
+
+fn multi_mode_gate_struct(ident: Ident) -> TokenStream {
+    quote! {
+        #[automatically_derived]
+        impl OperateMultiModeGate for #ident{
+            fn circuit(&self) -> crate::Circuit {
+                self.circuit
+            }
```

Comment:
I do not think we want to require a circuit decomposition here for the MultiModeGate